File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::unordered_map<int, int> freq;
        for (int num: nums) {
            freq[num]++;
        }

        std::vector<std::vector<int>> bucket(nums.size() + 1);
        for (const auto& [num, count]: freq) {
            bucket[count].push_back(num);
        }

        std::vector<int> res;
        for (int i = nums.size(); i > 0; i--) {
            for (int num: bucket[i]) {
                res.push_back(num);
                k--;
            }
            if (k == 0) {
                return res;
            }
        }
        
        return res;
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = Counter(nums)
        buckets = [[] for _ in range(len(nums) + 1)]
        for num, count in counter.items(): 
            buckets[count].append(num)
        
        res = []
        for i in range(len(nums), 0, -1):
            for num in buckets[i]:
                res.append(num)
                k -= 1
            if k == 0:
                return res
        
        return res